compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -53,15 +53,32 @@ pub enum SimplifiedType {
 pub enum TreatParams {
     /// Treat parameters as placeholders in the given environment.
     ///
-    /// Note that this also causes us to treat projections as if they were
-    /// placeholders. This is only correct if the given projection cannot
-    /// be normalized in the current context. Even if normalization fails,
-    /// it may still succeed later if the projection contains any inference
-    /// variables.
+    /// Note that this also causes us to treat projections (with no infer vars)
+    /// as if they were placeholders.
+    ///
+    /// This is only correct if the given projection cannot be normalized in
+    /// the current context. Even if normalization fails, it may still succeed
+    /// later if the projection contains any inference variables.
     AsPlaceholder,
+    // Same as above, but treats projections as placeholders even if they have
+    // infer vars. This is useful for and the new solver, which eagerly normalizes
+    // projections.
+    AsPlaceholderProjectionsAlways,
     AsInfer,
 }
 
+/// During fast-rejection, we have the choice of treating projection types with
+/// inference variables in its substitutions as placeholders or not.
+///
+/// In most cases, unless we know the type is fully normalized or we are handling
+/// the type separately, we should use "NotSimplifyable", as the type cannot be
+/// simplified as if it were rigid.
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum TreatProjectionsWithInfer {
+    AsPlaceholder,
+    NotSimplifyable,
+}
```

Comment:
I personally would prefer that this enum talks about the *why*, not the *what*. I think that makes it easier to choose which one to use.

```rust
pub enum TreatProjections {
    /// In the old solver we don't try to normalize projections
    /// when looking up impls and only access them by using the
    /// current self type. This means that if the self type is
    /// a projection which could later be normalized, we must not
    /// treat it as rigid.
    Default,
    /// We can treat projections in the self type as opaque as
    /// we separately look up impls for the normalized self type.
    TraitSolverNextSelfType,
}
```